File: 139-word-break/139.py
#!/usr/bin/env python
# -*- coding:utf-8 -*-

from typing import List


class Solution:
    def wordBreak(self, s: str, wordDict: List[str]) -> bool:
        size = len(s)
        dp = [False] * (size + 1)
        dp[0] = True
        for i in range(1, size + 1):
            for j in range(i):
                if dp[j] and s[j:i] in wordDict:
                    dp[i] = True
                    break

        return dp[-1]


if __name__ == '__main__':
    sol = Solution()
    s = "leetcode"
    wordDict = ["leet", "code"]
    print(sol.wordBreak(s, wordDict))

    s = "applepenapple"
    wordDict = ["apple", "pen"]
    print(sol.wordBreak(s, wordDict))

    s = "catsandog"
    wordDict = ["cats", "dog", "sand", "and", "cat"]
    print(sol.wordBreak(s, wordDict))

File: 139-word-break/139.cpp
#include <iostream>
#include <string>
#include <vector>
#include <unordered_set>

using namespace std;

class Solution {
public:
    bool wordBreak(string s, vector<string>& wordDict) {
        unordered_set<string> uset;
        for (auto& word : wordDict) {
            uset.insert(word);
        }
        int len = s.size();
        vector<int> dp = vector<int>(len+1, false);
        dp[0] = true;
        for (int i = 1; i < len+1; ++i) {
            for (int j = 0; j < i; ++j) {
                if (dp[j] && uset.count(s.substr(j, i-j))) {
                    dp[i] = true;
                    break;
                }
            }
        }
        return dp[len];
    }
};

int main(int argc, char** args) {
    Solution sol = Solution();
    string s1 = "leetcode";
    vector<string> wordDict1 = {"leet", "code"};
    cout << sol.wordBreak(s1, wordDict1) << endl;

    string s2 = "applepenapple";
    vector<string> wordDict2 = {"apple", "pen"};
    cout << sol.wordBreak(s2, wordDict2) << endl;

    string s3 = "catsandog";
    vector<string> wordDict3 = {"cats", "dog", "sand", "and", "cat"};
    cout << sol.wordBreak(s3, wordDict3) << endl;
    return 0;
}